src/frontend/src/repositories/anchorMetadata.ts

```diff
@@ -0,0 +1,194 @@
+type RawMetadataState = MetadataMapV2 | "loading" | "error" | "not-loaded";
+
+/**
+ * Class to manage the metadata of the anchor and interact with the canister.
+ *
+ * We decided to not throw any errors because this is non-critical for the application
+ * and we don't want to disrupt user flows if there is an error with the metadata.
+ *
+ * This class loads the metadata in the background when it's created.
+ * It can then be read and updated in memory.
+ * The metadata needs to be committed to the canister with the `commitMetadata` method to persist the changes.
+ */
+export class AnchorMetadataRepository {
+  // The nice AnchorMetadata is exposed to the outside world, while the raw metadata is kept private.
+  // We keep all the raw data to maintain other metadata fields.
+  private rawMetadata: RawMetadataState;
+  // Flag to keep track whether we need to commit the metadata to the canister.
+  private updatedMetadata: boolean;
+  private getter: MetadataGetter;
+  private setter: MetadataSetter;
+
+  static init = ({
+    getter,
+    setter,
+  }: {
+    getter: MetadataGetter;
+    setter: MetadataSetter;
+  }) => {
+    const instance = new AnchorMetadataRepository({ getter, setter });
+    // Load the metadata in the background.
+    void instance.loadMetadata();
+    return instance;
+  };
+
+  private constructor({
+    getter,
+    setter,
+  }: {
+    getter: MetadataGetter;
+    setter: MetadataSetter;
+  }) {
+    this.getter = getter;
+    this.setter = setter;
+    this.rawMetadata = "not-loaded";
+    this.updatedMetadata = false;
+  }
+
+  /**
+   * Load the metadata in the instance variable `rawMetadata`.
+   *
+   * This method won't throw an error if the metadata can't be loaded.
+   * Instead, it will set the instance variable `rawMetadata` to "error".
+   *
+   * @returns {Promise<void>} In case a client wants to wait for the metadata to be loaded.
+   */
+  loadMetadata = async (): Promise<void> => {
+    this.rawMetadata = "loading";
+    try {
+      this.updatedMetadata = false;
+      this.rawMetadata = await this.getter();
+    } catch (error) {
+      // Do not throw the error becasue this is not critical for the application.
+      this.rawMetadata = "error";
+      console.warn("Error loading metadata", error);
+      return;
+    }
+  };
+
+  private metadataIsLoaded = (
+    metadata: RawMetadataState
+  ): metadata is MetadataMapV2 => {
+    return (
+      this.rawMetadata !== "loading" &&
+      this.rawMetadata !== "error" &&
+      this.rawMetadata !== "not-loaded"
+    );
+  };
+
+  /**
+   * Returns the metadata transformed to `AnchorMetadata`.
+   *
+   * It tries to load the metadata if it's not loaded yet.
+   *
+   * It returns `undefined` if the metadata is not loaded.
+   *
+   * @returns {AnchorMetadata | undefined}
+   */
+  getMetadata = async (): Promise<AnchorMetadata | undefined> => {
+    if (!this.metadataIsLoaded(this.rawMetadata)) {
+      await this.loadMetadata();
```

Comment:
Shouldn't we simply await `this.rawMetadata` here? Otherwise you might load things twice / are actually implementing an implicit retry?

Given it's not critical, we should probably avoid that, no?

Reply:
As agreed, there won't be a `loadMetadata` call again and this function just returns the pretty metadata.